parser_library/src/context/using.cpp

```diff
@@ -0,0 +1,551 @@
+        });
+        u.resolve(*this, t);
+        u.compute_context(*this, t);
+    }
+}
+
+index_t<using_collection::instruction_context> using_collection::add(
+    dependency_evaluation_context ctx, processing_stack_t stack)
+{
+    m_instruction_contexts.push_back({ std::move(ctx), std::move(stack) });
+    return index_t<instruction_context>(m_instruction_contexts.size() - 1);
+}
+
+index_t<using_collection::mach_expression> using_collection::add(
+    std::unique_ptr<const mach_expression> expr, index_t<instruction_context> ctx)
+{
+    m_expr_values.push_back({ std::move(expr), ctx });
+    return index_t<mach_expression>(m_expr_values.size() - 1);
+}
+
+
+index_t<using_collection> using_collection::add(index_t<using_collection> current,
+    id_index label,
+    std::unique_ptr<mach_expression> begin,
+    std::unique_ptr<mach_expression> end,
+    std::vector<std::unique_ptr<mach_expression>> args,
+    dependency_evaluation_context eval_ctx,
+    processing_stack_t stack)
+{
+    assert(!args.empty() && args.size() <= reg_set_size);
+
+    index_t<instruction_context> ctx_id = add(std::move(eval_ctx), std::move(stack));
+
+    auto b = add(std::move(begin), ctx_id);
+    auto e = end ? add(std::move(end), ctx_id) : index_t<mach_expression>();
+
+    std::vector<index_t<mach_expression>> base;
+    base.reserve(args.size());
+
+    std::transform(args.begin(), args.end(), std::back_inserter(base), [this, ctx_id](auto& a) {
+        return add(std::move(a), ctx_id);
+    });
+
+    m_usings.emplace_back(current, ctx_id, b, base, label, e);
+    return index_t<using_collection>(m_usings.size() - 1);
+}
+
+index_t<using_collection> using_collection::remove(index_t<using_collection> current,
+    std::vector<std::unique_ptr<mach_expression>> args,
+    dependency_evaluation_context eval_ctx,
+    processing_stack_t stack)
+{
+    index_t<instruction_context> ctx_id = add(std::move(eval_ctx), std::move(stack));
+
+    std::vector<index_t<mach_expression>> base;
+    base.reserve(args.size());
+
+    std::transform(args.begin(), args.end(), std::back_inserter(base), [this, ctx_id](auto& a) {
+        return add(std::move(a), ctx_id);
+    });
+    m_usings.emplace_back(current, ctx_id, std::move(base));
+
+    return index_t<using_collection>(m_usings.size() - 1);
+}
+
+using_collection::evaluate_result using_collection::evaluate(
+    index_t<using_collection> context_id, id_index label, const section* owner, offset_t offset, bool long_offset) const
+{
+    if (!context_id)
+        return evaluate_result { invalid_register, 0 };
+
+    auto tmp = get(context_id).context.evaluate(label, owner, offset, long_offset);
+
+    if (tmp.length < 0)
+        return evaluate_result { invalid_register, 1 - tmp.length };
+    else
+        return evaluate_result { tmp.mapping_regs[0], tmp.reg_offset };
+}
+
+template</* std::integral */ typename R, /* std::integral */ typename T>
```

Comment:
Do you want to remove `/* std::integral */`?

Reply:
I would actually prefer to have it there... but clang-12 is not up to the task yet :(

Reply:
Ok